### PR TITLE
Dropterm

### PR DESCRIPTION
**Pull request by dmbates**
I think it would help to have functions for manipulating formulas.  `dropterm` is easy to define.  Methods could be added for a collection of `Symbol`s or expressions.  The reason I use this is to perform an analysis of deviance on a `GeneralizedLinearModel` (i.e. fit the original model, drop a term and fit the null model, evaluate the change in deviance).

### CODE REVIEW

File: src/formula.jl [hunk 1]
```diff
@@ -213,3 +213,23 @@ function Formula(t::Terms)
...
+
+Return a copy of `f` without the term `trm`.
+
+# Examples
+```jl
+julia> dropterm(foo ~ 1 + bar + baz, :bar)
+foo ~ 1 + baz
+```
+"""
+function dropterm(f::Formula, trm::Symbol)
```
Comment (kleinschmidt) [inline src/formula.jl]: Terms can be expressions, too (interactions terms). I think `Union{Expr,Symbol}` might do it here.

File: src/formula.jl [hunk 2]
```diff
@@ -213,3 +213,23 @@ function Formula(t::Terms)
...
+    rhs = copy(f.rhs)
+    args = rhs.args
+    if !(Meta.isexpr(rhs, :call) && args[1] == :+ && (tpos = findlast(args, trm)) > 0)
```
Comment (kleinschmidt) [inline src/formula.jl]: What about formulas with only one term, like `y ~ a`?

**Comment by ararslan** [comment]
It might be nice to provide a way to drop the intercept as well using this function. What would that look like? `dropterm(y ~ 1 + x, 1)`?

**Comment by ararslan** [comment]
Also, would it be worth it to provide `dropterm!` that modifies the `Formula` in place, with `dropterm(f, t) = dropterm!(copy(f), t)`?

File: src/formula.jl (hunk 2 above)
Comment (dmbates) [inline src/formula.jl]: I would like to discourage formulas like `y ~ a` and require either `y ~ 1 + a` or `y ~ 0 + a`

Also, what should `dropterm(y ~ a, :a)` return?  The lhs of a formula can be `nothing` but I don't know how to make sense of a rhs of `nothing`.

Comment (kleinschmidt) [inline src/formula.jl]: Yeah, that depends on the implicit intercept behavior. At the moment, I'd say it should return `y ~ 1` (e.g., the null model).

The `y ~ 0 + a` case is tricky, too; can you even evaluate `y ~ 0`?

I suppose you could just check whether the RHS would be empty and create `y ~ 1` for now, and we can fix it when we change intercepts to always be explicit.

Comment (ararslan) [inline src/formula.jl]: IMO `dropterm(y ~ a, :a)` should be an error

**Comment by dmbates** [comment]
@ararslan I agree that `dropterm!` should be the base definition and `dropterm` should call it.

Regarding dropping the intercept, that should be its own method (signature of `(::Formula, ::Integer)`) so that `dropterm(y ~ 1 + x, 1)` returns `y ~ 0 + x`, not `y ~ +x`.  An interesting question is what should `dropterm(y ~ 0 + x)` return?  It might return `y ~ 1 + x` or it might be cleaner to throw an error.

**Comment by ararslan** [comment]
> An interesting question is what should `dropterm(y ~ 0 + x)` return?

Are you missing an argument there? Do you mean `dropterm(y ~ 0 + x, 0)` or `dropterm(y ~ 0 + x, :x)`?

**Comment by dmbates** [comment]
@ararslan Yes, I meant "what should `dropterm(y ~ 0 + x, 0)` return?"

**Comment by ararslan** [comment]
Oh, okay. I would say that should be an error, since you're dropping a lack of intercept, which itself is a bizarre concept.

**Comment by dmbates** [comment]
@kleinschmidt @ararslan I have added a `dropterm!`, which `dropterm` now calls, and extended the signature of the argument to `trm::Union{Number, Symbol, Expr}`  I thinkt the CI failures for the nightly version are unrelated to these changes.

File: test/formula.jl [hunk 3]
```diff
@@ -106,4 +106,9 @@ t = Terms(y ~ x1 * x2 * x3)
 ## @test t.terms == [:x2, :(x1&x2)]    # == [:(1 & x1)]
 ## @test t.eterms == [:y, :x1, :x2]
 
+@test dropterm(foo ~ 1 + bar + baz, :bar).rhs == :(1 + baz)
+@test dropterm(foo ~ 1 + bar + baz, 1).rhs == :(0 + bar + baz)
+@test_throws ArgumentError dropterm(foo ~ 0 + bar + baz, 0)
+@test_throws ArgumentError dropterm(foo ~ 0 + bar + baz, :boz)
+
```
Comment (ararslan) [inline test/formula.jl]: Should test `dropterm!` as well. Maybe ensure that the input formula is indeed modified.

**Comment by ararslan** [comment]
I added a few commits to update this branch to accommodate the change to the `@formula` replacement for `Base.@~`. That should get the tests passing. LGTM, pending green CI.

**Comment by dmbates** [comment]
Thanks for updating this pull request @ararslan

**Comment by ararslan** [comment]
Hm, just realized that the docstring should have made its way into the documentation somewhere by adding it to one of the markdown files. Not really sure what the best section to put it in would be though.

**Comment by dmbates** [comment]
The `Formula` section of the `docs/src/formula.md`?

**Pull request merged**